### PR TITLE
nccl: 2.12.10-1 -> 2.16.5-1

### PR DESCRIPTION
**Pull request by ConnorBaker**
###### Description of changes

Version bump.

Passthru `cudaPackages` instead of `cudatoolkit`. No users in Nixpkgs currently so nothing breaks.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

###### Failures

- [x] tensorflow
  - Unrelated and reproducible on `master`.
  - https://gist.github.com/ConnorBaker/3ab689a988e0bb157b7312ae33bf33a5 

###### Future work

- Refactor to use `cudaPackages` instead of `cudatoolkit` and set `NVCC_GENCODE` using `cudaFlags`.
- Refactor to use `autoAddOpenGLRunpathHook`
- Refactor to remove dependency on `cudatoolkit`

cc @NixOS/cuda-maintainers

### CODE REVIEW

**Comment by ConnorBaker** [comment]
After doing

```bash
gh pr checkout 215552 \
    && gh pr checkout 215549 \
    && git checkout master \
    && git merge cudaGencode-fix --no-edit \
    && git merge nccl-2.16.5-1 --no-edit \
    && nixpkgs-review rev master
```

I got

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>cudaPackages.nccl</li>
    <li>python310Packages.cupy</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.tensorflowWithCuda</li>
    <li>python310Packages.torchWithCuda</li>
    <li>python311Packages.cupy</li>
    <li>python311Packages.jaxlibWithCuda</li>
  </ul>
</details>

**Comment by ConnorBaker** [comment]
Result of `nixpkgs-review pr 215549` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>python310Packages.tensorflowWithCuda</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>cudaPackages.nccl</li>
    <li>python310Packages.cupy</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.torchWithCuda</li>
    <li>python311Packages.cupy</li>
    <li>python311Packages.jaxlibWithCuda</li>
  </ul>
</details>

File: pkgs/development/libraries/science/math/nccl/default.nix [hunk 1]
```diff
@@ -1,14 +1,18 @@
...
   name = "nccl-${version}-cuda-${cudatoolkit.majorVersion}";
-  version = "2.12.10-1";
+  version = "2.16.5-1";
 
   src = fetchFromGitHub {
     owner = "NVIDIA";
     repo = "nccl";
     rev = "v${version}";
-    sha256 = "sha256-QqORzm0gD+QG+P8rId8bQn2oZsxL5YyxCIobUVs85wE=";
+    sha256 = "sha256-JyhhYKSVIqUKIbC1rCJozPT1IrIyRLGrTjdPjJqsYaU=";
```
Comment (samuela) [inline pkgs/development/libraries/science/math/nccl/default.nix:15]:
```suggestion
    hash = "sha256-JyhhYKSVIqUKIbC1rCJozPT1IrIyRLGrTjdPjJqsYaU=";
```

**Review by samuela — COMMENTED** [review]
Thanks for putting this together @ConnorBaker ! just one small nit

**Pull request merged**